route-snapper/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+use geom::{Circle, Distance, FindClosest, LonLat, PolyLine, Pt2D};
+use map_model::{IntersectionID, Map, PathConstraints, RoadID};
+use wasm_bindgen::prelude::*;
+
+const INTERSECTON_RADIUS: Distance = Distance::const_meters(10.0);
+
+#[wasm_bindgen]
+pub struct JsRouteSnapper {
+    map: Map,
+    snap_to_intersections: FindClosest<IntersectionID>,
+    route: Route,
+    mode: Mode,
+}
+
+#[derive(Clone, PartialEq)]
+struct Route {
+    waypoints: Vec<IntersectionID>,
+    full_path: Vec<IntersectionID>,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+enum Mode {
+    Neutral,
+    Hovering(IntersectionID),
+    Dragging { idx: usize, at: IntersectionID },
+}
+
+#[wasm_bindgen]
+impl JsRouteSnapper {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Result<JsRouteSnapper, JsValue> {
+        // Panics shouldn't happen, but if they do, console.log them.
+        console_error_panic_hook::set_once();
+
+        let map: Map = abstutil::from_binary(include_bytes!("../bristol.bin"))
+            .map_err(|err| JsValue::from_str(&err.to_string()))?;
+
+        let mut snap_to_intersections = FindClosest::new(map.get_bounds());
+        for i in map.all_intersections() {
+            snap_to_intersections.add_polygon(i.id, &i.polygon);
+        }
+
+        Ok(Self {
+            map,
+            snap_to_intersections,
+            route: Route::new(),
+            mode: Mode::Neutral,
+        })
+    }
+
+    #[wasm_bindgen(js_name = renderGeojson)]
+    pub fn render_geojson(&self) -> String {
+        let mut pairs = Vec::new();
+
+        let gps_bounds = Some(self.map.get_gps_bounds());
+
+        // Draw the confirmed route
+        for pair in self.route.full_path.windows(2) {
+            // TODO Inefficient!
+            let r = self
+                .map
+                .get_r(self.map.find_road_between(pair[0], pair[1]).unwrap());
+            pairs.push((
+                r.center_pts.to_geojson(gps_bounds),
+                make_props("type", "confirmed route"),
+            ));
+        }
+        for i in &self.route.full_path {
+            pairs.push((
+                Circle::new(self.map.get_i(*i).polygon.center(), INTERSECTON_RADIUS)
```

Comment:
Passing back tessellated circles is probably silly, we could just send back a single point and use maplibre styling to draw it nicely.